compatibility_spark-3/src/main/scala/za/co/absa/hyperdrive/compatibility/impl/writer/cdc/hudi/scd2/HudiCDCToSCD2Writer.scala

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.compatibility.impl.writer.cdc.hudi.scd2
+
+import org.apache.commons.configuration2.Configuration
+import org.apache.commons.lang3.StringUtils
+import org.apache.hudi.DataSourceWriteOptions._
+import org.apache.hudi.QuickstartUtils.getQuickstartWriteConfigs
+import org.apache.hudi.exception.TableNotFoundException
+import org.apache.spark.sql.streaming.{OutputMode, StreamingQuery, Trigger}
+import org.apache.spark.sql.{DataFrame, DataFrameWriter, Row, SaveMode}
+import org.slf4j.LoggerFactory
+import za.co.absa.hyperdrive.compatibility.impl.writer.cdc.CDCUtil.{SCD2Fields, getSchemaWithSCD2Fields, getStagedDataForSCD2, isDirEmptyOrDoesNotExist}
+import za.co.absa.hyperdrive.ingestor.api.utils.{ConfigUtils, StreamWriterUtil}
+import za.co.absa.hyperdrive.ingestor.api.writer.{StreamWriter, StreamWriterFactory}
+
+import java.io.FileNotFoundException
+
+private[writer] class HudiCDCToSCD2Writer(destination: String,
+                                          trigger: Trigger,
+                                          checkpointLocation: String,
+                                          partitionColumns: Seq[String],
+                                          keyColumn: String,
+                                          timestampColumn: String,
+                                          operationColumn: String,
+                                          operationDeleteValues: Seq[String],
+                                          precombineColumns: Seq[String],
+                                          precombineColumnsCustomOrder: Map[String, Seq[String]],
+                                          val extraConfOptions: Map[String, String]) extends StreamWriter {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  private val CheckpointLocation = "checkpointLocation"
+  private val HoodieCommitTime = "_hoodie_commit_time"
+  private val HoodieCommitSeqno = "_hoodie_commit_seqno"
+  private val HoodieRecordKey = "_hoodie_record_key"
+  private val HoodiePartitionPath = "_hoodie_partition_path"
+  private val HoodieFileName = "_hoodie_file_name"
+
+  if (StringUtils.isBlank(destination)) {
+    throw new IllegalArgumentException("Destination must not be blank!")
+  }
+
+  override def write(dataFrame: DataFrame): StreamingQuery = {
+    dataFrame.writeStream
+      .trigger(trigger)
+      .outputMode(OutputMode.Append())
+      .option(CheckpointLocation, checkpointLocation)
+      .options(extraConfOptions)
+      .foreachBatch((input: DataFrame, batchId: Long) => {
+        logger.info(s"Writing batchId: $batchId")
+        val spark = input.sparkSession
+        val dataFrameSchema = getSchemaWithSCD2Fields(dataFrame)
+
+        var hudiTable: DataFrame = null
```

Comment:
I would extract this to a separate function `createHudiTableIfNotExists`

Reply:
Done